be/src/olap/rowset/segment_v2/segment_iterator.cpp

```diff
@@ -2836,125 +2401,46 @@ Status SegmentIterator::current_block_row_locations(std::vector<RowLocation>* bl
+        _common_expr_ctxs_push_down.emplace_back(context);
     }
+    return Status::OK();
 }
 
-void SegmentIterator::_calculate_func_in_remaining_conjunct_root() {
-    auto hash = [](const vectorized::VExprSPtr& expr) -> std::size_t {
-        return std::hash<std::string>()(expr->expr_name());
-    };
-    auto equal = [](const vectorized::VExprSPtr& lhs, const vectorized::VExprSPtr& rhs) -> bool {
-        return lhs->equals(*rhs);
-    };
-
-    uint32_t next_id = 0;
-    std::unordered_map<vectorized::VExprSPtr, uint32_t, decltype(hash), decltype(equal)> unique_map(
-            0, hash, equal);
-
-    auto gen_func_unique_id = [&unique_map, &next_id](const vectorized::VExprSPtr& expr) {
-        auto it = unique_map.find(expr);
-        if (it != unique_map.end()) {
-            return it->second;
-        } else {
-            unique_map[expr] = ++next_id;
-            return next_id;
-        }
-    };
-
+void SegmentIterator::_calculate_expr_in_remaining_conjunct_root() {
     for (const auto& root_expr_ctx : _common_expr_ctxs_push_down) {
         const auto& root_expr = root_expr_ctx->root();
         if (root_expr == nullptr) {
             continue;
         }
 
-        std::stack<std::pair<vectorized::VExprSPtr, bool>> stack;
-        stack.emplace(root_expr, false);
+        std::stack<vectorized::VExprSPtr> stack;
+        stack.emplace(root_expr);
 
         while (!stack.empty()) {
-            const auto& [expr, has_compound_pred] = stack.top();
+            const auto& expr = stack.top();
             stack.pop();
 
-            bool current_has_compound_pred =
-                    has_compound_pred || (expr->node_type() == TExprNodeType::COMPOUND_PRED);
-
-            if (expr->fn().name.function_name == "multi_match") {
-                expr->set_index_unique_id(gen_func_unique_id(expr));
-                if (current_has_compound_pred) {
-                    compound_func_exprs.emplace_back(expr);
-                } else {
-                    no_compound_func_exprs.emplace_back(expr);
-                }
-
-                for (int32_t i = expr->get_num_children() - 1; i >= 0; i--) {
-                    auto child_expr = expr->get_child(i);
-                    if (child_expr->node_type() == TExprNodeType::type::SLOT_REF) {
-                        std::string result_sign = BeConsts::BLOCK_TEMP_COLUMN_PREFIX +
-                                                  std::to_string(expr->index_unique_id());
-                        _func_name_to_result_sign[child_expr->expr_name()].push_back(result_sign);
+            if (vectorized::VExpr::is_directly_action_on_a_slot(*expr)) {
+                for (const auto& child : expr->children()) {
+                    if (child->is_slot_ref()) {
+                        auto* column_slot_ref = assert_cast<vectorized::VSlotRef*>(child.get());
+                        _common_expr_inverted_index_status[column_slot_ref->expr_name()]
```

Comment:
`expr_name` in VSlotRef corresponds to `column_name`. You might wonder why not use `column_id` instead—this is because the `column_id` in the expr can differ from the one in `segment_iterator`.